codersdk/deploymentconfig.go

```diff
@@ -12,77 +12,163 @@ import (
+	AllowSignups      *DeploymentConfigField[bool]     `json:"allow_signups" typescript:",notnull"`
+	EnterpriseBaseURL *DeploymentConfigField[string]   `json:"enterprise_base_url" typescript:",notnull"`
+}
+
+type OIDCConfig struct {
+	AllowSignups *DeploymentConfigField[bool]     `json:"allow_signups" typescript:",notnull"`
+	ClientID     *DeploymentConfigField[string]   `json:"client_id" typescript:",notnull"`
+	ClientSecret *DeploymentConfigField[string]   `json:"client_secret" typescript:",notnull"`
+	EmailDomain  *DeploymentConfigField[string]   `json:"email_domain" typescript:",notnull"`
+	IssuerURL    *DeploymentConfigField[string]   `json:"issuer_url" typescript:",notnull"`
+	Scopes       *DeploymentConfigField[[]string] `json:"scopes" typescript:",notnull"`
+}
+
+type TelemetryConfig struct {
+	Enable *DeploymentConfigField[bool]   `json:"enable" typescript:",notnull"`
+	Trace  *DeploymentConfigField[bool]   `json:"trace" typescript:",notnull"`
+	URL    *DeploymentConfigField[string] `json:"url" typescript:",notnull"`
+}
+
+type TLSConfig struct {
+	Enable       *DeploymentConfigField[bool]     `json:"enable" typescript:",notnull"`
+	CertFiles    *DeploymentConfigField[[]string] `json:"cert_file" typescript:",notnull"`
+	ClientAuth   *DeploymentConfigField[string]   `json:"client_auth" typescript:",notnull"`
+	ClientCAFile *DeploymentConfigField[string]   `json:"client_ca_file" typescript:",notnull"`
+	KeyFiles     *DeploymentConfigField[[]string] `json:"key_file" typescript:",notnull"`
+	MinVersion   *DeploymentConfigField[string]   `json:"min_version" typescript:",notnull"`
 }
 
 type Flaggable interface {
 	string | bool | int | time.Duration | []string
 }
 
 type DeploymentConfigField[T Flaggable] struct {
-	Key string `json:"key"`
-	// Name appears in the deployment UI.
 	Name       string `json:"name"`
 	Usage      string `json:"usage"`
 	Flag       string `json:"flag"`
 	Shorthand  string `json:"shorthand"`
 	Enterprise bool   `json:"enterprise"`
 	Hidden     bool   `json:"hidden"`
+	Secret     bool   `json:"secret"`
+	Default    T      `json:"default"`
 	Value      T      `json:"value"`
 }
 
+// MarshalJSON removes the Value field from the JSON output of any fields marked Secret.
+// nolint:revive
+func (f *DeploymentConfigField[T]) MarshalJSON() ([]byte, error) {
+	if !f.Secret {
+		return json.Marshal(struct {
+			Name       string `json:"name"`
+			Usage      string `json:"usage"`
+			Flag       string `json:"flag"`
+			Shorthand  string `json:"shorthand"`
+			Enterprise bool   `json:"enterprise"`
+			Hidden     bool   `json:"hidden"`
+			Secret     bool   `json:"secret"`
+			Default    T      `json:"default"`
+			Value      T      `json:"value"`
+		}{
+			Name:       f.Name,
+			Usage:      f.Usage,
+			Flag:       f.Flag,
+			Shorthand:  f.Shorthand,
+			Enterprise: f.Enterprise,
+			Hidden:     f.Hidden,
+			Secret:     f.Secret,
+			Default:    f.Default,
+			Value:      f.Value,
+		})
```

Comment:
Could we just marshal ourselves in this case?

```
json.Marshal(f)
```

Reply:
You can't! I did that first and you get an infinite loop because it just calls this marshaled again. LOL